subsys/bluetooth/host/gatt.c

```diff
@@ -2674,6 +2671,17 @@ uint8_t bt_gatt_check_perm(struct bt_conn *conn, const struct bt_gatt_attr *attr
 	}
 
 	mask &= attr->perm;
+
+	if (mask & BT_GATT_PERM_LESC_MASK) {
+#if defined(CONFIG_BT_SMP)
+		if ((conn->le.keys->flags & BT_KEYS_SC) == 0) {
+			return BT_ATT_ERR_AUTHENTICATION;
+		}
+#else
+		return BT_ATT_ERR_INSUFFICIENT_ENCRYPTION;
```

Comment:
I think this one shouldn't be changed. In this case we dont have the SMP, meaning there is no pairing.

Reply:
In the authentication check
```c
	if (mask & BT_GATT_PERM_AUTHEN_MASK) {
		if (bt_conn_get_security(conn) < BT_SECURITY_L3) {
			return BT_ATT_ERR_AUTHENTICATION;
		}
	}
```
we return `BT_ATT_ERR_AUTHENTICATION` regardless of `CONFIG_BT_SMP`.

I expect the reason why the encryption check
```
	if ((mask & BT_GATT_PERM_ENCRYPT_MASK)) {
#if defined(CONFIG_BT_SMP)
		if (!conn->encrypt) {
			return BT_ATT_ERR_INSUFFICIENT_ENCRYPTION;
		}
#else
		return BT_ATT_ERR_INSUFFICIENT_ENCRYPTION;
#endif /* CONFIG_BT_SMP */
	}
```
use it, is because `conn->encrypt` is guarded by `CONFIG_BT_SMP`:
```c
#if defined(CONFIG_BT_SMP) || defined(CONFIG_BT_BREDR)
	bt_security_t		sec_level;
	bt_security_t		required_sec_level;
	uint8_t			encrypt;
#endif /* CONFIG_BT_SMP || CONFIG_BT_BREDR */
```

So I do think that we should return the same error message regardless of whether `CONFIG_BT_SMP` is enabled or not.

Reply:
I think that in the latter case, with encryption, you will end up in a situation without encryption in both cases, which is why I think its correct to return `BT_ATT_ERR_INSUFFICIENT_ENCRYPTION`.

I'm not sure about the authenticated case though, if its correct or a miss.

Regardless, I do think that we actually want to do something like this: 

```
	if (mask & BT_GATT_PERM_LESC_MASK) {
#if defined(CONFIG_BT_SMP)
		if ((conn->le.keys->flags & BT_KEYS_SC) == 0 &&
		     conn->encrypt) {
			return BT_ATT_ERR_AUTHENTICATION;
		}
#endif /* CONFIG_BT_SMP */
		return BT_ATT_ERR_INSUFFICIENT_ENCRYPTION;
	}
```

Reply:
BT_ATT_ERR_INSUFFICIENT_ENCRYPTION is for case when encryption is not enabled and we have LTK
for any other case BT_ATT_ERR_AUTHENTICATION should be used

Reply:
> I think that in the latter case, with encryption, you will end up in a situation without encryption in both cases, which is why I think its correct to return `BT_ATT_ERR_INSUFFICIENT_ENCRYPTION`.
> 
> I'm not sure about the authenticated case though, if its correct or a miss.
> 
> Regardless, I do think that we actually want to do something like this:
> 
> ```
> 	if (mask & BT_GATT_PERM_LESC_MASK) {
> #if defined(CONFIG_BT_SMP)
> 		if ((conn->le.keys->flags & BT_KEYS_SC) == 0 &&
> 		     conn->encrypt) {
> 			return BT_ATT_ERR_AUTHENTICATION;
> 		}
> #endif /* CONFIG_BT_SMP */
> 		return BT_ATT_ERR_INSUFFICIENT_ENCRYPTION;
> 	}
> ```

@sjanc in the case above, correct me if I'm wrong, when we do not have CONFIG_BT_SMP enabled, this means encryption can't be enabled, and we are right to return BT_ATT_ERR_INSUFFICIENT_ENCRYPTION.